BT-Fetch

resession opened the pull request:
pull request to bt-fetch

resession committed:
make changes

resession commented:
"I think having users configure something in their .agregorerc file in the btOptions section."

currently trying to finish up the user configuration. tried to research on .agregorerc, but i'm still not seeing an answer. how do i access the properties/contents of .agregorerc?

resession committed:
make changes

resession committed:
make changes

resession committed:
make changes

resession committed:
make changes

resession committed:
make changes

resession committed:
make changes

resession committed:
make changes

resession committed:
make changes

resession committed:
make changes

resession committed:
make changes

resession committed:
make changes

resession committed:
make changes

resession committed:
make changes

resession committed:
make changes

resession committed:
make changes

resession committed:
make changes

resession commented:
Here are are the 2 packages that this pull request relies on.

https://github.com/resession/bt-fetch-torrent
https://github.com/resession/bt-fetch-property

RangerMauve commented:
Thank you for putting so much effort into this. However it might take me a while to review it all because I'm having trouble understanding the code flow and formatting. In particular I'm not sure what [bt-fetch-torrent](https://github.com/resession/bt-fetch-torrent/blob/main/index.js) since it has a lot of code within it and doesn't have a README. 😅

I'll try to spend a good chunk of time tonight to look through everything. If it's okay, I might do a bit of rewriting to the PR to make it easier for me to maintain going forward.

RangerMauve commented on index.js:
```diff
@@ -1,291 +1,450 @@
+                        } else {
+                            if(body.folder){
+                                if(body.address === undefined || body.secret === undefined){
+                                    let {torrent, title} = await app.publishTitle(body.folder)
```
I don't think we can realistically include code like this that would expose users filesystem to arbitrarily being put into a torrent and available on the internet. 🤔 It's a major security concern and would be way too easy to abuse by malicious code.

Maybe there could be a way to prompt users to choose a folder to upload instead? Might be good to bikeshed this concept a bit. If we can't handle FormData parsing right from the stream into the torrent, it might be safe to load the data into a temporary folder and then create the torrent from that.

IMO the ideal is still to use FormData to upload a set of files into a torrent directly.

Before then it might be safe to provide a way to publish a mutable torrent from an infohash? Then it'd be up to users to create the torrent through some other means. 🤔

resession replied:
got it, i'll table it for tomorrow and get it done.

resession committed:
make changes

resession commented:
wanted to finish up a basic version and then optimize later as things come up. i know there might be a lot to go through, so i will go ahead and create a readme for all modules.

resession commented:
feel free to make pull requests or any changes. i am currently making the readme files. I'll also keep making changes. I have a change that will make things lighter and easier to maintain.

resession committed:
make changes

resession commented:
i changed the protocol scheme to `bt`, it's shorter and easy. i see that bt-fetch used to use `bt` but it got changed to `bittorrent`. so if you wanna change it back to bittorrent than that's fine.

resession committed:
make changes

resession committed:
make changes

resession commented:
little background: i tried making this torrent fetch module twice. this is my second time. at the first time, there were too many things happening at the same time. fetch + managing torrents + bep46 + removing old data + updating torrents. so i chose to split the fetch part from everything else. so that is why i made sub-modules(bt-fetch-torrent and bt-fetch-property). this second time, the fetch module was a breeze.

so here is the current change that i made. i combined and merged bt-fetch-torrent and bt-fetch-property(the sub-modules) into one module. also, instead of having bt-fetch-torrent/bt-fetch-property as npm modules you have to install, i just combined them into a single file called 'main.js' in the bt-fetch repo. that way if something needs to be changed, its right there, so it can easily be changed.

i am still going through things to see what optimizations and improvements i can make. since webtorrent as of right now does not return duplicate torrents, some functions are very similar or almost duplicates, but one handles publishing new torrents and the other handles a torrent that was alreeady published(i will see how i can make this better, that way have less code). i'll post again after my next updates.

resession committed:
make changes

resession committed:
make changes

resession commented:
made a lot of comments on the main.js which is now a part of the pull request. it explains what each function does.

resession committed:
make changes

resession committed:
make changes

resession committed:
make changes

resession committed:
make changes

resession committed:
make changes

resession committed:
make changes

resession committed:
make changes

resession committed:
make changes

resession committed:
make changes

resession committed:
make changes

resession committed:
make changes

resession committed:
make changes

resession committed:
make changes

resession committed:
make changes

resession committed:
make changes

RangerMauve commented on index.js:
```diff
@@ -1,291 +1,430 @@
+                                res.headers['X-Downloaded'] = `${tempData.downloaded}`
+                                res.statusCode = 200
+                            } else {
+                                let foundFile = tempData.files.find(data => {return data.path.endsWith(req.mainPath)})
```
```suggestion
                                let foundFile = tempData.files.find(data => data.path.endsWith(req.mainPath))
```

resession replied:
fixed

RangerMauve commented on index.js:
```diff
@@ -1,291 +1,430 @@
+                case 'GET': {
+                    if(req.mainType){
+                        if(req.mainReq){
+                            res.data = ['<html><head><title>BT-Fetch</title></head><body><div><p>timeout: ' + app._status.timeout + '</p><p>Torrents: ' + app.webtorrent.torrents.length + '</p><p>initial: ' + app._status.initial + '</p><p>current: ' + app._status.current + '</p><p>share: ' + app._status.share + '</p><div></body></html>']
```
Would it be possible to use a template literal here instead?

resession replied:
fixed

RangerMauve commented on index.js:
```diff
@@ -1,291 +1,430 @@
+                            res.data = [JSON.stringify({timeout: app._timeOut, share: app._status.share, current: app._status.current, initial: app._status.initial, torrents: app.webtorrent.torrents.length})]
+                        }
+                        res.statusCode = 200
+                        res.headers['Content-Type'] = req.mainRes
```
Would it be possible to add some strict CORS settings so that random websites don't have access to this info?

resession replied:
fixed, its better not to return this type of information, the user who is using bt-fetch should already know the options beforehand since they are the ones running the module. so i took this part out.

RangerMauve commented on index.js:
```diff
@@ -1,291 +1,430 @@
-      if (!infoHash || !infoHash.match(INFO_HASH_MATCH)) {
-        throw new Error('Infohash must be 40 char hex string')
-      }
+              if((protocol !== 'bt:' || !method || !SUPPORTED_METHODS.includes(method) || !hostname) || (hostname.length !== 1 && hostname.length !== 32 && hostname.length !== 40 && hostname.length !== 64) || (hostname.length === 1 && hostname !== hostType) || (hostname.length !== 1 && !checkTitle.test(hostname) && !checkHash.test(hostname) && !checkAddress.test(hostname))){
```
Could you split this up into individual if statements with different error messages? Ideally this would result in status codes that make sense for what the error is (e.g. 409 for invalid methods) This should help folks debug what's wrong.

Also, mind switching the protocol back to `bittorrent`? I know it's more verbose, but there's also past precedence for using this in older browsers. 😁

resession replied:
fixed

RangerMauve commented on index.js:
```diff
@@ -1,291 +1,430 @@
-          const contentType = wantsHTML ? 'text/html; charset=utf-8' : 'application/json; charset=utf-8'
-          headers['Content-Type'] = contentType
-          // TODO: Account for folders
+              let req = formatReq(hostname, pathname, method, searchParams, reqHeaders)
```
Would it make sense to take the logic inside of `formatReq` and yank it outside of the function / have it at the top instead? That way when you're scanning the code you can see what the variables mean right away without having to dart down to the bottom.

resession replied:
fixed

RangerMauve commented on main.js:
```diff
@@ -0,0 +1,910 @@
+        finalOpts.folder = path.resolve(finalOpts.folder)
+        fs.ensureDirSync(finalOpts.folder)
+
+        this._status = {current: finalOpts.current, share: finalOpts.share, timeout: finalOpts.timeout, initial: finalOpts.initial}
```
Could these properties be placed directly on `Main`?

resession replied:
fixed

RangerMauve commented on main.js:
```diff
@@ -0,0 +1,910 @@
+        this._readyToGo = true
+
+        // run the start up function
+        startUp(this).catch(error => {console.log(error)})
```
Could this be refactored to something like `this.startUp().catch()` so that all the methods are actual methods?

resession replied:
fixed

RangerMauve commented on main.js:
```diff
@@ -0,0 +1,910 @@
+        startUp(this).catch(error => {console.log(error)})
+
+        // run the keepUpdated function every 1 hour, it keep the data active by putting the data back into the dht, don't run it if it is still working from the last time it ran the keepUpdated function
+        setInterval(() => {
```
The return value from this `setInteval` should be saved as a variable so that we can call `clearInterval` when shutting down the process.

resession replied:
fixed

RangerMauve commented on main.js:
```diff
@@ -0,0 +1,910 @@
+        }
+        let data = await fs.readFile(this._author + path.sep + address)
+        data = JSON.parse(data.toString())
+        if(infoHash !== data.infoHash || !ed.verify(Buffer.from(data.sig, 'hex'), encodeSigData({seq: data.sequence, v: {ih: infoHash, ...data.stuff}}), Buffer.from(data.address, 'hex'))){
```
Could the variables in this if statement be split up a bit to improve readability? e.g. `matchesSignature = ed.verify(whatever..`, `signature = Buffer.from(data.sig, 'hex')`, `encodedSignatureData = encodeSigData(etc...`

resession replied:
fixed

RangerMauve commented on main.js:
```diff
@@ -0,0 +1,910 @@
+                    reject(new Error(address + ' property took too long, it timed out, please try again with only the keypair without the folder'))
+                },this._status.timeout)
+            }),
+            new Promise((resolve, reject) => {
```
I don't think you need to use `new Promise()` here. Would be better to put this code into a method and have `await this.checkProperty()`

Maybe the `webtorrent.remove()` could be placed into a `catch` block and the timeout could just reject?

resession replied:
i will need elaboration on this to make sure i understand. will table this until i get other stuff done.

RangerMauve commented on main.js:
```diff
@@ -0,0 +1,910 @@
+            return haveTorrent
+        }
+        // if user had torrent before, then empty the folder so previous data does not conflict with the infohash
+        let folderPath = this._external + path.sep + hash
```
Could you use `path.join(this._external, hash)` instead of using `path.sep`?

resession replied:
fixed

RangerMauve commented on main.js:
```diff
@@ -0,0 +1,910 @@
+        //     console.log(error)
+        // }
+        checkTorrent = await Promise.race([
+            new Promise((resolve, reject) => {
```
Might be useful to create a function for timeouts. e.g. Also, it might be a good time to import the [delay module](https://www.npmjs.com/package/delay)

`await rejectAfter(interval, new Error('some message'))`

Could look something like this:

```
async function rejectAfter(interval, error) {
  await delay(interval)
  throw error
}
```

resession replied:
i will get to this later, need some elaboration.

RangerMauve replied:
Mainly, there's a lot of duplicate code for "reject with an error after a certain timeout", the function I proposed could clean up a lot of it and make the code more readable.

e.g. this code could look like

```JavaScript
checkTorrent = await Promise.race([
      rejectAfter(this.timeout, new Error(hash + ' took too long, it timed out'))
      new Promise((resolve, reject) => {
        this.webtorrent.add(hash, { path: folderPath, destroyStoreOnDestroy: true }, torrent => {
          resolve(torrent)
        })
      })
    ])
```

That said, I'm seeing `webtorrent.add()` invoke d a bunch of times with promises wrappers, so it might be good to add a method for that too.

e.g.

```JavaScript
async addToWebTorrent(...args) {
    return new Promise((resolve) => {
    this.webtorrent.add(...args, resolve)
     })
}
```

which could make the final code look like this:

```JavaScript
checkTorrent = await Promise.race([
      rejectAfter(this.timeout, new Error(hash + ' took too long, it timed out'))
      this.addToWebTorrent(hash, { path: folderPath, destroyStoreOnDestroy: true })
    ])
```

Which I think will be a lot easier to read and maintain going forward. 😁

RangerMauve commented on main.js:
```diff
@@ -0,0 +1,910 @@
+        ])
+        checkTorrent.folder = folderPath
+        checkTorrent.side = false
+        checkTorrent.title = crypto.createHash('md5').update(hash).digest("hex")
```
What is the purpose of hashing the hash in order to get the title?

resession replied:
i have that there just to be consistent with the publishTitle function. also a little ocd lol. but i'll take that out.

resession replied:
fixed

RangerMauve commented on index.js:
```diff
@@ -1,291 +1,430 @@
+                            if(prog.has(req.mainQuery)){
+                                tempData = prog.get(req.mainQuery)
+                            } else {
+                                tempData = await app.loadAddress(req.mainQuery)
```
Cuould `tempData` be renamed to `torrentData` to better reflect what it represents?

resession replied:
fixed

RangerMauve commented on index.js:
```diff
@@ -1,291 +1,430 @@
+                            }
+                            if(req.mainPath === path.sep){
+                                if(req.mainReq){
+                                    res.data = [`<html><head><title>${tempData.address}</title></head><body><div>${tempData.files.map(file => {return `<p><a href="bt://${tempData.address}/${file.path.replace(tempPath, '').split(path.sep).map(data => {return encodeURIComponent(data)}).join('/')}">${file.name}</a></p>`})}</div></body></html>`]
```
Could we use relative URLs here? Also the `bittorrent:` protocol?

resession replied:
fixed

RangerMauve commented on index.js:
```diff
@@ -1,291 +1,430 @@
+                                if(req.mainReq){
+                                    res.data = [`<html><head><title>${tempData.address}</title></head><body><div>${tempData.files.map(file => {return `<p><a href="bt://${tempData.address}/${file.path.replace(tempPath, '').split(path.sep).map(data => {return encodeURIComponent(data)}).join('/')}">${file.name}</a></p>`})}</div></body></html>`]
+                                } else {
+                                    res.data = [JSON.stringify(tempData.files.map(file => {return 'bt://' + tempData.address + '/' + file.path.replace(tempPath, '').split(path.sep).map(data => {return encodeURIComponent(data)}).join('/')}))]
```
Could this have relative paths the way that the previous version of bt-fetch had it?

resession replied:
fixed

RangerMauve commented on index.js:
```diff
@@ -1,291 +1,430 @@
+                                res.headers['Content-Type'] = req.mainRes
+                                res.headers['Content-Length'] = tempData.length
+                            } else {
+                                foundFile = tempData.files.find(file => {return file.path.endsWith(req.mainPath)})
```
Could this be done with `file.path === req.mainPath` instead? `endsWith` could lead to bugs.

resession replied:
fixed

RangerMauve commented on index.js:
```diff
@@ -1,291 +1,430 @@
+                                        if(ranges && ranges.length && ranges.type === 'bytes'){
+                                            let [{ start, end }] = ranges
+                                            let length = (end - start + 1)
+                                            req.mainPartial.start = start
```
Is the `req.mainPartial` variable necessary here? Could we not just use `{start, end}` inside the `createReadStream` API?

resession replied:
fixed

RangerMauve commented on index.js:
```diff
@@ -1,291 +1,430 @@
+                                    }
+                                } else {
+                                    res.data = [JSON.stringify('file was not found')]
+                                    res.statusCode = 400
```
Not Found errors are `404`. Might be good to be consistent with [the spec](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status#client_error_responses)

resession replied:
fixed

RangerMauve commented on index.js:
```diff
@@ -1,291 +1,430 @@
+                                    }
+                                    res.statusCode = 200
+                                } else {
+                                    let {torrent, secret} = await app.publishAddress(body.folder, {address: body.address, secret: body.secret})
```
Yeah, instead of publishing a folder, it'd be good to be able to specify a infohash to publish, unless we can figure out the FormData thing. 😅

I like your idea of specifying the secret key within the request, it should give applications more control over stuff. I'll see if I can adapt this approach to IPNS. Though it might need to go into a searchParams or in the headers for cases where we upload data in the request body. 🤔 Or maybe it _could_ just be a parameter? Might need to think about it more.

resession replied:
publishing the folder would mean specifying the folder and then we copy the data to the _internal directory and then seed it. but sure, we can go the formData route as well, also more secure. that should work fine. this is something i will get to later after i finish some of the other stuff.

RangerMauve replied:
Here's the code I'm using to parse out FormData https://github.com/RangerMauve/js-ipfs-fetch/blob/default/index.js#L101

Likely you'll want to create a tmp folder to upload the files, then create the torrent, from it, then create a folder for the torrents infohash, then close the torrent and copy the data over, then reopen the torrent. 😅

RangerMauve commented on index.js:
```diff
@@ -1,291 +1,430 @@
+                        }
+                        res.headers['Content-Type'] = req.mainRes
+                    } else {
+                        if(req.mainQuery.length === 64){
```
Could you elaborate on what this request means? When you `POST` to a public key it fetches the latest version? Would it be possible to use `GET` for this instead? POST is usually useful when uploading data somewhere.

resession replied:
this is when a user has stopped a previous torrent that was running, and then they want to restart that same torrent.

RangerMauve replied:
Would it be possible to fold that functionality into doing a `GET` on the regular bittorrent URL with the torrent info instead?

resession replied:
`GET` is used to load torrents of other user's, meaning `webtorrent.add()`. i will have to think about this and see how it can be done in the best way.

RangerMauve commented on index.js:
```diff
@@ -1,291 +1,430 @@
+                            res.statusCode = 400
+                            // mainData = [await app.clearData()]
+                            // prog.clear()
+                        } else if(body.hash){
```
Rather than specifying the hash in the body, would it make sense to do a DELETE on the hash in the hostname? e.g. `DELETE bittorrent://infohash_here/` This would also play well with having the public key in the URL.

resession replied:
the DELETE method already handles that right now. if a user does `DELETE bittorrent://infohash_here/`, then we stop the torrent only, if a user wants to stop the torrent and remove the data from disk then they would need to put the infohash in the body. what are your thoughts, do you want to change it?

RangerMauve replied:
Oh dang. Stopping to seed without removing the data seems like a useful case, but I haven't done anything like that before. 🤔

One precedent from the `hypercore-fetch` repo was to have special paths under the `$` folder, like `hyper:92798237492374/$/extesnions/`, would it make sense to standardize around a `/$/seeding` file where we can GET to see the status of whether it's seeding and `POST` to update it's settings? Could be useful to bring that to other protocol handlers too. Having stuff related to specific torrents be under `_` still feels a bit weird though.

Any thoughts? 😅

RangerMauve commented on index.js:
```diff
@@ -1,291 +1,430 @@
 
-      const headers = {}
+        if(request.body !== null){
+            request.body = await getBody(request.body)
```
Would it make sense to only load the body like this where you actually need it? This will conflict with any code that does stream based processing of the body like FormData

resession replied:
made an if statement, so when a user is using the `DELETE` or `POST` and the hostname is `_` then we do the following

`if(body !== null){
          body = await getBody(body)
          try {
            body = JSON.parse(body)
          } catch (error) {
            console.log(error)
          }
        }`

RangerMauve commented on main.js:
```diff
@@ -0,0 +1,910 @@
+    // if share option is true, then start seeding(webtorrent.add seeds after torrent is downloaded) on start up
+    if(self._status.share){
+        let checkExternal = await fs.readdir(self._external, {withFileTypes: false})
+        for(let i = 0;i < checkExternal.length;i++){
```
Use `for(const externalPath of checkExternal)` style for loops when possible.

resession replied:
fixed

RangerMauve commented on main.js:
```diff
@@ -0,0 +1,910 @@
+            let folderPath = self._external + path.sep + checkExternal[i]
+            if(checkAddress.test(checkExternal[i])){
+                let checkProperty = await Promise.any([
+                    new Promise((resolve, reject) => {setTimeout(() => {resolve(null)}, self._status.timeout)}),
```
This could be replaced with the `delay` module

resession replied:
i will work on this

resession replied:
do you think it would be good to import a module just for the delay feature since we would be bringing in another module for only 1 thing?

RangerMauve replied:
I mean, I usually just use the `delay` module, but defining it as a utility function is super easy too.

```
async function delay(time) {
  return new Promise((resolve)=> setTimeout(resolve, time))
}
```

resession replied:
made the `delayTimeOut()` function, will that satisfy what we are looking for?

RangerMauve replied:
Should be good! Might be useful to set a default for the last argument (e.g. instead of `delayTimeOut(timeout, data, res)` have `delayTimeOut(timeout, data, shouldResolve = false)`

RangerMauve commented on main.js:
```diff
@@ -0,0 +1,910 @@
+                            try {
+                                await fs.emptyDir(folderPath)
+                            } catch (error) {
+                                console.log(error)
```
In which cases can this fail?

resession replied:
followed the fs-extra docs. this is how they have it when `await` is used. i do not see this ever failing. it creates the directory if it does not exist. but their documentation has it inside a try block so i thought maybe thet is safer since it might throw something 0.01% of the time. i'll considered this fixed since we are following their docs, if you disagree, let me know anytime.

https://github.com/jprichardson/node-fs-extra/blob/HEAD/docs/emptyDir.md

RangerMauve replied:
Personally I try to let errors bubble up whenever possible, but if you're set on this, it'd be good to use `console.error` instead of `console.log` since STDERR is more appropriate for error messages.

RangerMauve commented on main.js:
```diff
@@ -0,0 +1,910 @@
+
+        // if current option is true, then if the infohash for the address is brand new then empty the directory and download the new infohash
+        // if the current option is false, then at least make sure the main folder which is named with the public key address exists
+        if(this._status.current){
```
Do we really need to empty the directory first? I thought that webtorrent would be able to reuse any files that hand't changed.

resession replied:
it empties the directory only if current infohash is not found(meaning if the current infohash for the pubkey has not been downloaded before), which is exactly what the `current` option is. we only keep the the most updated infohash. if the infohash has not been changed, then the directory will not be be emptied. if the `current` option is off/falsy, then we just make sure the pubkey directory exists and that's it. this should be fixed, but if you have other thoughts, let me know.

RangerMauve commented on main.js:
```diff
@@ -0,0 +1,910 @@
+
+    // publish a new BEP46 torrent, or update an address/public key with a new torrent
+    // we return secret key along with the torrent because the user will need it
+    async publishAddress(folder, keypair){
```
Instead of a folder, it'd be useful to be able to just update the infoHash.

resession replied:
originally i wanted BEP46 torrents to be used for sites. like building a vuejs app and then torrenting the `dist` folder where the app resides. but for security reasons, i understand that torrenting the folder might not be the best thing to do. so by updating the infohash i assume you mean update the current torrent by using formData. sure, i will work on that.

RangerMauve commented on main.js:
```diff
@@ -0,0 +1,910 @@
+          sha1(buffAddKey, (targetID) => {
+      
+            this.webtorrent.dht.get(targetID, (getErr, getData) => {
+              if(getErr){
```
Should we be rejecting when encountering errors here?

resession replied:
i don't think we should reject for the get error. something might not broadcast anymore but we still also have the data ourselves. so if something doesn't broadcast(hence the get error), we just put it back into the dht. so we reject if we can not put the data back. which is what is currently happening.

resession replied:
if you are okay with that, i will consider this fixed.

RangerMauve replied:
Yup that makes sense!

RangerMauve reviewed (changes requested):
Hey! Thanks for all the progress. I've reviewed some code and left some comments. In addition to the stuff I already said, would you mind running the code through the [standard.js](https://standardjs.com/) linter to keep the code style consistent with the rest of the codebase?

I think the main TODO is to figure out the folder stuff more, I don't think it's safe to expose arbitrary folders to be turned into torrents like that, and having FormData for uploads seems like the safer bet.

I'm also unsure about how the `title` thing is supposed to work and whether there's another approach we could take there. Is this supposed to be similar to the "pet names" feture in ipfs-fetch and hypercore-fetch?

For uploading I think it'd be good to have the following:
- Being able to `POST` a `bittorrent://` URL represented by a public key with an infohash in order to update it
- Being able to `POST` some FormData to a `bittorrent://` URL represented by a public key in order to create a new torrent from the FormData files (similar to how IPFS and hypercore do it) and publish its infohash.
- Being able to `POST` some FormData to a `bittorrent://` URL with an existing infohash which will take the existing torrent data, then add files from your FormData to it (updating existing ones), returning a bittorrent:// URL with the new infohash
-  Being able to `DELETE` a file path within an bittorrent URL, if it's an existing infohash, deleting that file or folder from the torrent, if it's a public key, fetching the torrent for that public key and deleting the file and then updating the public key.

Another thing I'm a bit worried about is all of the interaction with the filesystem. It might be worth it to use a single [store](https://github.com/webtorrent/webtorrent/blob/master/lib/torrent.js#L80) between all of the torrents so that data could be aggressively deduplicated. It might be good to create on at the top level and pass it down everywhere. I'm not sure how we'll know which chunks of the store should be cleared out though.

But yeah, thanks again for working on this!I think we're getting a lot closer to making the bittorrent code catch up with IPFS and hypercore. 😁

resession committed:
make changes

resession committed:
make changes

resession committed:
make changes

resession committed:
make changes

resession committed:
make changes

resession commented:
made lots of changes. main one is adding formdata. most of what you brought up has been added and resolved. i will be testing the changes and then we can see if the changes are satisfactory.

resession committed:
make changes

resession committed:
make changes

resession committed:
make changes

resession committed:
make changes

resession committed:
make changes

RangerMauve commented:
Sweet, thank you for the updates! I'm probably going to need to look into this more over the weekend since I've got a bunch on my plate during the week again. 😭

resession committed:
make changes

resession committed:
make changes

resession committed:
make changes

resession commented:
awesome, yeah check it out whenever you get some time. i am waiting until everything is ready. once we are ready, then i will convert it to standardjs style.

resession committed:
make changes

resession committed:
make changes

resession committed:
make changes

resession committed:
make changes

resession committed:
make changes

resession committed:
make changes

resession committed:
make changes

resession committed:
make changes

resession committed:
make changes

resession committed:
make changes

resession committed:
make changes

resession committed:
make changes

RangerMauve commented on index.js:
```diff
@@ -1,291 +1,311 @@
+                  res.headers['X-Downloaded'] = `${foundFile.downloaded}`
+                  res.statusCode = 200
+                } else {
+                  res.statusCode = 400
```
This should be `404` for "not found"

RangerMauve commented on main.js:
```diff
@@ -0,0 +1,889 @@
+    if(hash){
+      this.stopHash(hash)
+    } else {
+      hash = crypto.createHash('sha1').update(crypto.randomBytes(20).toString('hex')).digest('hex')
```
The hashing shouldn't be necessary here if randomBytes is being used. Alternatly, it might be useful to create a tmp dir to upload the torrent data to (in order to get an infohash) and then creating the folder from the infohash and moving the torrent data over.

RangerMauve commented on main.js:
```diff
@@ -0,0 +1,889 @@
+      function handleFiles(name, file, info){
+        // fs.writeFile(path.join(folderPath, info.filename), Readable.from(file))
+        // const saveTo = fs.createWriteStream(path.join(folderPath, info.filename))
+        file.pipe(fs.createWriteStream(path.join(folderPath, info.filename)))
```
Try wrapping `file` in `Readable.from()`, the internal busboy file streams might be doing something wonky.

RangerMauve commented on main.js:
```diff
@@ -0,0 +1,889 @@
+      }
+      bb.on('file', handleFiles)
+      bb.on('error', handleErrors)
+      bb.on('close', handleClose)
```
Try listening on the `finish` event instead, this is [what I have in js-ipfs-fetch](https://github.com/RangerMauve/js-ipfs-fetch/blob/default/index.js#L105)

RangerMauve reviewed (changes requested):
Hopefully some of these comments will help with the form-data issues you were having. 😅

resession commented:
"This should be 404 for "not found""
fixed.

"The hashing shouldn't be necessary here if randomBytes is being used. Alternatly, it might be useful to create a tmp dir to upload the torrent data to (in order to get an infohash) and then creating the folder from the infohash and moving the torrent data over."
webtorrent takes the name of the directory into account when it comes to the infohash of the torrent, so if we name of the new directory is the infohash we get from the tempdir, then the torrent's infohash of the new directory(after moving the data) will be different).

"Try wrapping file in Readable.from(), the internal busboy file streams might be doing something wonky."
fixed.

"Try listening on the finish event instead, this is what I have in js-ipfs-fetch"
fixed.

resession committed:
make changes

RangerMauve commented:
> webtorrent takes the name of the directory into account when it comes to the infohash of the torrent, so if we name of the new directory is the infohash we get from the tempdir, then the torrent's infohash of the new directory(after moving the data) will be different).

Is there a way to override that functionality? What about requiring that people specify a name for the torrent when they create one?

resession commented:
> > webtorrent takes the name of the directory into account when it comes to the infohash of the torrent, so if we name of the new directory is the infohash we get from the tempdir, then the torrent's infohash of the new directory(after moving the data) will be different).
> 
> Is there a way to override that functionality? What about requiring that people specify a name for the torrent when they create one?

unfortunately, i do not see a way to override this aspect of webtorrent. we might be able to work with the torrent name, i'll maike some changes and see how it goes.

resession commented:
i looked at it and played around, tried torrent names. yes we can require users to give us a name for the torrent/dir. here is the comparison.

if we go the name route, then the fetch part has to change to handle general names on the hostname portion of the request(protocol://hostname/path) instead of `bittorrent://someHashOrKey`, we now would have to handle `bittorrent://someNameOfTorrent`, this would be true for the non-BEP46 torrent we are seeding

if we keep it as it is, then what happens is when we seed a non-BEP46 torrent, we create a random sha1 hash that is not the infohash but more of an identifier for non-BEP46 torrents we seed(when we leech, we torrent.hash = torrent.infoHash, so for leech hash is the infohash, not for seed though).

my thoughts is that the second way is more simple. i like the bittorrent://40CharHashOr64CharKey, so this would be better in my opinion, simpler, and feels bittorrenty since its hashes. i would go with this and this is how we currently have it.

while we would like to have something directly connected to the torrent we seed like the torrent name or infohash, but we would have to add things that might cancel out the benefit(changing fetch, adding things).

i do not mind if you want to go the first route though, if you wanna go the first route then we can do that.

------------------------

i thought about maybe extending the first route. if you want to use torrent names then you can extend that to everything since the names would be more friendlier than hashes. basically when we GET bittorrent://someHashOrKey, instead of Map.set('hashOrKey'), we Map.set('nameOfTorrent'). also then bittorrent links would have to be encoded. we can dictate rules like not having spaces for torrent names on agregore, but torrents that is created else where would have to have there names url encoded that way we can create valid bittorrent://torrentNameAsHostname links. this actually sounds pretty cool because then people would only need hashes and keys when they first visit a torrent but after that they can just use regular names for the torrents in the url. we can keep things simple now and maybe we can test this and add in if it makes sense.

edit: you can ignore the block of text above this. i just realized that this would go highly against standards and would not make a lot of sense. read it for thought exercise purpose but this would not be fetchy and therefore, most likely not doable. using an infohash/key for the first request and then using the torrent name(general string) for the next requests for the same torrent isn't something people will be used to.

---------------------

so as far as having hashes vs using torrent name for seeding a regular torrent, what are your thoughts?

resession committed:
make changes

resession committed:
make changes

resession committed:
make changes

resession commented:
"Try wrapping file in Readable.from(), the internal busboy file streams might be doing something wonky."

by the way, i did this. crossed my fingers. unfortunately, it didn't fix the issue. i'll think and test more.

resession committed:
make changes

resession committed:
make changes

resession committed:
make changes

resession commented:
"Also, mind switching the protocol back to `bittorrent`? I know it's more verbose, but there's also past precedence for using this in older browsers."

can you elaborate on past precendence on old browsers? do you mean older versions of agregore, or web browsers in general. i'm not aware of any browsers that handle `bittorrent`://, do you mean when web browsers has to use a custom protocol to run an external app(like when it asks you to select a default app for something)?

RangerMauve commented:
Regarding `bt:` vs `bittorrent:`, I've been discussing this for a while in [this thread](https://github.com/bittorrent/bittorrent.org/issues/92)

tl;dr

- There was a browser called [Maelstrom](https://www.ctrl.blog/entry/bittorrent-maelstrom.html) which used `bittorrent://` (pointed out in [this thread](https://twitter.com/RangerMauve)
- It was proposed [in the ietf mailing list](https://mailarchive.ietf.org/arch/msg/p2prg/D6utRAOKTxahrQKwSEIWy4wqPrE)
- I think it's pretty obvious from seeing the URL scheme what it's fore, bt (although not that much differ) is a bit less obvious.

resession commented:
thanks for the thread, went over it.

i heard about maelstorm before, i think it got abandoned or never took off(something to due with investors or something the last time i read about it). the more important part is the ietf mailing list, but my counter point is that it is proposed, it is not fully accepted and not fully standardized yet.

you are right, `bittorrent` is more obvious, but i think people's comfort zone is more important. `http` is not obvious, but its quick and simple. less typing, the better for the average user, `bt` is really quick. this isn't high on my list or as important, but i am thinking from the standpoint of the average user. more along the lines of user adoption. whatever makes it quick, simple and easy for them, i am for. my opinion is use `bt` right now, if `bittorrent` becomes accepted and standardized then we can make the changes when that happens. but again, this is not high on my personal list, if you prefer `bittorrent` and want to keep it that way, i am fine with that.

resession committed:
use standardjs

RangerMauve commented:
Gonna close this in favor of #3 🎉
